kernel/src/engine/arrow_expression/evaluate_expression.rs

```diff
@@ -477,15 +492,97 @@ pub fn to_json(input: &dyn Datum) -> Result<ArrayRef, ArrowError> {
     }
 }
 
+/// Coalesce multiple arrays into one by selecting the first non-null value from each row.
+///
+/// This function implements SQL COALESCE semantics: for each row, it iterates through
+/// the input arrays from left to right and returns the first non-null value found. If all values
+/// are null for a given row, the result will be null for that row.
+///
+/// # Parameters
+/// - `arrays`: Slice of Arrow arrays to coalesce. Must not be empty and all arrays must have the same data type.
+/// - `result_type`: Optional expected result type. If provided, must match the arrays' data type.
+///
+/// # Returns
+/// An `ArrayRef` containing the coalesced values with the same number of rows as the input arrays.
+///
+/// # Errors
+/// This function returns an `ArrowError` in the following cases:
+/// - **Empty input**: The default engine currently does not support empty COALESCE statements.
+/// - **Mismatched row counts**: Not all arrays have the same number of rows.
+/// - **Mismatched data types**: Not all arrays have exactly the same data type.
+/// - **Invalid result type**: If `result_type` is provided but doesn't match the arrays' data type.
+pub fn coalesce_arrays(
```

Comment:
does it need to be `pub`? it's really just powering `COALESCE` inside expression eval right?

Reply:
If we wanted to benchmark it in the expression_bench at a later point, it would need to be `pub`, right?

I can make `coalesce_arrays()` private for now since we don't use it in benchmarks yet, but `to_json()` (same context, different expression) already is `pub` and used outside the crate in our expression benchmark.

Reply:
I don't think its worth blocking on but I'd prefer to keep pub(crate) or make `#[internal_api]` if we need? can make a follow up to normalize decision with to_json side as well

Reply:
Agreed. I created https://github.com/delta-io/delta-kernel-rs/issues/1316 to track this.